## Carregar dados.

In [1]:

import sys
import os
sys.path.append(os.path.abspath(".."))

from utils.limpeza import carregar_dados

# Carregar dataset
df = carregar_dados()

# Verificar colunas disponíveis
print("Colunas disponíveis:")
print(df.columns.tolist())

# Visualizar tipos e nulos
df.info()
df.isna().mean().sort_values(ascending=False).head(20)



[INFO] Dados carregados com sucesso: 5217 linhas e 403 colunas.
Colunas disponíveis:
['0.a_token', '0.d_data/hora_envio', '1.a_idade', '1.a.1_faixa_idade', '1.b_genero', '1.c_cor/raca/etnia', '1.d_pcd', '1.e_experiencia_profissional_prejudicada', '1.e.1_Não acredito que minha experiência profissional seja afetada', '1.e.2_Sim, devido a minha Cor/Raça/Etnia', '1.e.3_Sim, devido a minha identidade de gênero', '1.e.4_Sim, devido ao fato de ser PCD', '1.i.1_uf_onde_mora', '1.f.1_Quantidade de oportunidades de emprego/vagas recebidas', '1.f.2_Senioridade das vagas recebidas em relação à sua experiência', '1.f.3_Aprovação em processos seletivos/entrevistas', '1.f.4_Oportunidades de progressão de carreira', '1.f.5_Velocidade de progressão de carreira', '1.f.6_Nível de cobrança no trabalho/Stress no trabalho', '1.f.7_Atenção dada pelas pessoas diante das minhas opiniões e ideias', '1.f.8_Relação com outras pessoas da empresa, em momentos de trabalho', '1.f.9_Relação com outras pessoas da empre

1.h_pais_onde_mora                                                                                                   0.973356
5.b_oportunidade_buscada                                                                                             0.945179
5.d_experiencia_em_processos_seletivos                                                                               0.944604
5.c_tempo_em_busca_de_oportunidade                                                                                   0.944412
5.a_objetivo_na_area_de_dados                                                                                        0.896301
6.g_ferramentas_de_qualidade_de_dados_(dia_a_dia)                                                                    0.892659
6.f_tecnologia_data_warehouse                                                                                        0.860840
6.d_tecnologia_data_lake                                                                                             0

# Força o recarregamento.

In [60]:
import importlib
import sys

# Remove o módulo da memória
if 'app.recomendador' in sys.modules:
    del sys.modules['app.recomendador']

# Força o recarregamento
from app.recomendador import RecomendadorRH
importlib.reload(sys.modules['app.recomendador'])

<module 'app.recomendador' from 'c:\\Users\\Tiago\\Desktop\\rh\\app\\recomendador.py'>

- Garante que a versão mais atualizada de app.recomendador esteja carregada, útil durante o desenvolvimento, especialmente em notebooks, onde o Python não recarrega automaticamente módulos alterados.

## Teste do Filtro
- cargo geral: Cargo Atual + Cargo desejado.
- cargo atual: 2.f_cargo_atual
- cargo desejado: 5.b_oportunidade_buscada
- senioridade: 2.g_nivel
- experiencia dados: 2.i_tempo_de_experiencia_em_dados
- experiencia_ti: 2.j_tempo_de_experiencia_em_ti
- modelo_trabalho_atual: 2.r_modelo_de_trabalho_atual
- modelo_trabalho_ideal: 2.s_modelo_de_trabalho_ideal
- faixa salarial: 2.h_faixa_salarial
- uf_onde_mora: 1.i.1_uf_onde_mora
- regiao_onde_mora: 1.i.2_regiao_onde_mora
- satisfação: 2.k_satisfeito_atualmente

In [56]:
#teste

from app.recomendador import RecomendadorRH
from utils.limpeza import carregar_dados

df = carregar_dados()

recom = RecomendadorRH(df)

criterios = {
    "cargo_geral": "cientista"
    "",
    "senioridade": "Sênior",
    "linguagens_programacao": ["Python"],
    "clouds": ["Google Cloud"]
}


df_resultado = recom.aplicar_filtros(criterios)
print(f"Candidatos encontrados: {recom.contar_candidatos()}")
recom.top_n(5)


[INFO] Dados carregados com sucesso: 5217 linhas e 403 colunas.

[LOG] Aplicando filtros com interseção (AND):
✅ Cargo geral 'cientista': 913 matches
🔍 Processando filtro simples: senioridade
✅ 1573 matches em 2.g_nivel
🔧 Processando múltiplos valores para linguagens_programacao
✅ 2935 matches em 4.d.3_Python
🔧 Processando múltiplos valores para clouds
✅ 1050 matches em 4.h.2_Google Cloud (GCP)

🎯 Total de candidatos qualificados (todos critérios): 128
Candidatos encontrados: 128


0.a_token  0.d_data/hora_envio  1.a_idade  \
524  4l3dsshwehb298rsqq54l3dsh5y681jx  13/11/2024 21:15:30         24   
680  gwe6m9xmcwcbr86p2gwe6mcgqmqdmw23  29/10/2024 14:17:49         25   
693  mpobzdhd721cpuhbb1s6i4mpobzdlpkn  05/11/2024 20:31:12         25   
750  falqn5avjmeqv12rjmfalqn5amw4ppmv  28/11/2024 16:14:59         25   
826  wdrnzd7toobbnamywdrnp8zi2ns10i2x  14/10/2024 12:28:43         26   

    1.a.1_faixa_idade 1.b_genero 1.c_cor/raca/etnia 1.d_pcd  \
524             22-24  Masculino             Branca     Não   
680             25-29  Masculino             Branca     Não   
693             25-29  Masculino             Branca     Não   
750             25-29  Masculino             Branca     Não   
826             25-29  Masculino             Branca     Não   

    1.e_experiencia_profissional_prejudicada  \
524                                      NaN   
680                                      NaN   
693                                      NaN   
750                                      NaN   
826                                      NaN   

     1.e.1_Não acredito que minha experiência profissional seja afetada  \
524                                                NaN                    
680                                                NaN                    
693                                                NaN                    
750                                                NaN                    
826                                                NaN                    

     1.e.2_Sim, devido a minha Cor/Raça/Etnia  ...  \
524                                       NaN  ...   
680                                       NaN  ...   
693                                       NaN  ...   
750                                       NaN  ...   
826                                       NaN  ...   

     8.d.3_Entrando em contato com os times de negócio para definição do problema, identificar a solução e apresentação de resultados.  \
524                                                0.0                                                                                   
680                                                NaN                                                                                   
693                                                0.0                                                                                   
750                                                0.0                                                                                   
826                                                0.0                                                                                   

     8.d.4_Desenvolvendo modelos de Machine Learning com o objetivo de colocar em produção em sistemas (produtos de dados).  \
524                                                0.0                                                                        
680                                                NaN                                                                        
693                                                1.0                                                                        
750                                                1.0                                                                        
826                                                1.0                                                                        

    8.d.5_Colocando modelos em produção, criando os pipelines de dados, APIs de consumo e monitoramento.  \
524                                                0.0                                                     
680                                                NaN                                                     
693                                                0.0                                                     
750                                                0.0                                                     
826          

# VISÃO GERAL

### Distribuição por senioridade dos profissionais.
- Sênior reperesentam o maior grupo (1573 representantes), seguidos por Plenos (1377) e Júniores (868)
- Isso mostra que a concentração de profissionais experientes é maior o que pode indiciar maturidade na área de dados, ou que profissionais mais experientes são mais propensos a responder à pesquisa.

In [11]:
# @title Distribuição por Senioridade
import plotly.express as px

fig = px.histogram(
    df,
    x="2.g_nivel",
    title="Distribuição por Senioridade dos Profissionais",
    labels={"2.g_nivel": "Nível"},
    color="2.g_nivel",
    text_auto=True
)
fig.update_layout(xaxis_title="Nível", yaxis_title="Contagem")
fig.show()


### Distribuição Regional dos profissionais
- Sudeste domina com 60% dos profissionais.
- Em seguida o sul (19,8%), Nordeste (9,89%), Centro-Oeste (6,35%) e Norte (2,78%).
- Esses dados mostram a desigualdade regional no setor de tecnologia e dados. Regiãoes que tem maior desenvolvimento econômico concentram mais profissionais da área. 
- Pode haver oportunidades para desccentralizar profissionais com trabalho remoto. 

In [ ]:
# @title Distribuição por região do Brasil
fig = px.pie(
    df,
    names="1.i.2_regiao_onde_mora",
    title="Distribuição Regional dos Profissionais",
    hole=0.4
)
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()


### Distribuição das faixas salariais.
- As faixas de R$ 8.000 a R$ 12.000 e de R$ 12.000 a R$ 16.000 tem o maior número de profissionais.
- Salários mais altos de (R$ 20.000 a 25.000) e mais baixos (menor que R$ 2.000) são menos presentes.
- Essa faixa de salário observada indica uma concentração de nível intermediário e alto, o que condiz com o núemro elevao de profissionais sêniors.

In [61]:
# @title Faixa Salarial
fig = px.histogram(
    df,
    y="2.h_faixa_salarial",
    title="Distribuição das Faixas Salariais",
    labels={"2.h_faixa_salarial": "Faixa Salarial"},
    color="2.h_faixa_salarial",
    text_auto=True,
    orientation="h"
)
fig.update_layout(yaxis={'categoryorder':'total ascending'})
fig.show()


### Tempo de experência em dados por senioridade.
- Existe um crescimento evidente no tempo de experência conforme a senioridade.
- Júnior: É predominante até 2 anos.
- Pleno: Já é mais intermediário, entre 3 e 6 anos.
- Sênior: Maioria acima de 5 anos, muitos com mais de 7 anos. 
- Essa distribuição é esperada com a progressão na carreira.

In [62]:
# @title Tempo de Experiência por Senioridade 

import plotly.express as px

fig = px.box(df, 
             x="2.g_nivel", 
             y="2.i_tempo_de_experiencia_em_dados", 
             color="2.g_nivel",
             color_discrete_sequence=px.colors.sequential.Blues,
             title="Tempo de Experiência em Dados por Senioridade",
             labels={
                 "2.g_nivel": "Nível",
                 "2.i_tempo_de_experiencia_em_dados": "Tempo de Experiência (anos)"
             })

# Personalização das bordas e estilo
fig.update_traces(
    marker=dict(
        line=dict(width=1, color='gray')  # Bordas mais finas e cinza
    ),
    line=dict(width=1.5, color='gray')    # Linhas medianas/whiskers
)

# Ajustes de layout
fig.update_layout(
    xaxis={'categoryorder':'total descending'},
    showlegend=False,
    plot_bgcolor='white',
    yaxis=dict(gridcolor='lightgray'),
    xaxis_tickangle=-15,
    margin=dict(l=20, r=20, t=40, b=20)  # Margens mais compactas
)

fig.show()

# Machine Learning

### Aplicar clustering em profissionais da área de dados com abase em experiência e stack técnica (linguagens e cloud).
- A clusterização foi feita com base em variáveis técnicas, como anos de experiência em TI  e Dados, linguagens (Python, SQL, R etc) e nuvens (AWS, GCP, Azure).
- O uso de PCA permite reduzir de 10 variávies para 2 componentes princiapis para visualização.
- Foi escolhido n_clusters=4 foi torna adequada a segmentação de perfis distintos. E o cluster_tecnico marca a segmentação de perfis técnicos.

In [14]:
# @title Clusterização de perfis técnicos com KMeans
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA

# Mapeamento de faixas de tempo para valores aproximados em anos
mapa_experiencia = {
    "Menos de 1 ano": 0.5,
    "de 1 a 2 anos": 1.5,
    "de 2 a 3 anos": 2.5,
    "de 3 a 4 anos": 3.5,
    "de 4 a 5 anos": 4.5,
    "de 5 a 6 anos": 5.5,
    "de 6 a 7 anos": 6.5,
    "de 7 a 8 anos": 7.5,
    "de 8 a 9 anos": 8.5,
    "de 9 a 10 anos": 9.5,
    "Mais de 10 anos": 12
}

# Aplica o mapeamento
df["experiencia_dados_anos"] = df["2.i_tempo_de_experiencia_em_dados"].map(mapa_experiencia)
df["experiencia_ti_anos"] = df["2.j_tempo_de_experiencia_em_ti"].map(mapa_experiencia)

# Colunas para ML
colunas_ml = [
    "experiencia_dados_anos",
    "experiencia_ti_anos",
    "4.d.1_SQL",
    "4.d.2_R",
    "4.d.3_Python",
    "4.d.6_Java",
    "4.d.14_JavaScript",
    "4.h.1_Amazon Web Services (AWS)",
    "4.h.2_Google Cloud (GCP)",
    "4.h.3_Azure (Microsoft)"
]

# Prepara o DataFrame para clustering
df_ml = df[colunas_ml].copy().fillna(0)

# Normaliza os dados
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df_ml)

# Redução de dimensionalidade para visualização
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_scaled)

# Aplica KMeans
kmeans = KMeans(n_clusters=4, random_state=42)
df["cluster_tecnico"] = kmeans.fit_predict(X_scaled)



### Interpretação dos componentes principais (PCA).
- Componente 1 (Eixo X) está associado a linguagens e expêriencia técnica, sendo influenciado fortemente por: SQL, Python, Java, JavaScript, e AWS. Esse eixo pode ser interpretado como uma medida de senioridade e stack técnica mais robusta. 
- Componente 2 (Eixo Y) está ligado a linguagens alternativas e diversidade na stack, destaques em: JavaScript, Java, e Azure. Pode ser iterpretada como um indicador de stack mais variada e perfil híbrido que é "menos tradicional" do posto de vista técnico.

In [25]:
# Interpretação das Componentes Principais (eixos X e Y)
import pandas as pd

# Cria DataFrame com os pesos (loadings) de cada variável nas componentes
componentes = pd.DataFrame(
    pca.components_,
    columns=colunas_ml,
    index=["Componente 1", "Componente 2"]
).T

# Ordena por importância absoluta em cada componente
componentes["abs_comp1"] = componentes["Componente 1"].abs()
componentes["abs_comp2"] = componentes["Componente 2"].abs()

# Mostra as variáveis que mais influenciam cada eixo
print("🔹 Top variáveis que definem o EIXO X (Componente Principal 1):")
display(componentes.sort_values("abs_comp1", ascending=False)[["Componente 1"]].head(5))

print("🔸 Top variáveis que definem o EIXO Y (Componente Principal 2):")
display(componentes.sort_values("abs_comp2", ascending=False)[["Componente 2"]].head(5))


🔹 Top variáveis que definem o EIXO X (Componente Principal 1):


Componente 1
4.d.1_SQL                            0.518025
4.d.3_Python                         0.507364
4.d.6_Java                           0.349382
4.h.1_Amazon Web Services (AWS)      0.344215
4.d.14_JavaScript                    0.330140

🔸 Top variáveis que definem o EIXO Y (Componente Principal 2):


Componente 2
4.d.14_JavaScript                    0.610043
4.d.6_Java                           0.596815
4.d.3_Python                        -0.276120
4.d.1_SQL                           -0.262357
4.h.1_Amazon Web Services (AWS)     -0.200809

### Heatmap - Experiência média por cluster x senioridade.
- Mesmo entre profissionais com a mesma senioridade declarada, a média de experiência em dados varia bastante clusters. Por exemplo profissionais sêniores no cluster 2 têm média de 5 anos de experiência em dados, enquanto no cluster 1 apenas 3,3 anos. 
- Isso revela que a senioridade percebida no mercado não necessariamente reflete o tempo efetivo de experiência técnica. 
- O cluster 2 destaca-se como o grupo com maior tempo de experiência média entre todas as senioridades. 

In [27]:
# @title visualização dos clusters
import plotly.express as px

fig = px.scatter(
    x=X_pca[:, 0],
    y=X_pca[:, 1],
    color=df["cluster_tecnico"].astype(str),
    title="Clusters Técnicos com KMeans",
    labels={"color": "Cluster"},
    hover_data=[df["2.f_cargo_atual"], df["2.g_nivel"]]
)

fig.update_layout(xaxis_title="Experiência + Linguagens Técnicas(Componente 1)", yaxis_title="Cloud & Stack Variada(componente 2)")
fig.show()


### Distribuição de Senioridade por Cluster.
- O cluster 0 tem maior concentração de profissionais ( princiaplmente sêniores).
- O cluster 3 tem uma boa representatividad em todas as senioridades, especilmanete plenos e sêniores.
- O cluster 1 possui mais juniores, indicando possível grupo de entrada na área ou transição. 

| Cluster | Interpretação                                                | Perfil técnico dominante                                      | Senioridade dominante                           |
| ------- | ------------------------------------------------------------ | ------------------------------------------------------------- | ----------------------------------------------- |
| **0**   | Generalistas de nível alto                                   | Stack tradicional (SQL), uso forte de AWS, experiência sólida | **Sênior** (maioria), seguido de Pleno          |
| **1**   | Profissionais menos técnicos ou generalistas de outras áreas | Pouca proficiência técnica, "Outra opção" como cargo          | **Júnior e Pleno**                              |
| **2**   | Especialistas em Cloud e Dados                               | Experientes, alto uso de cloud, destaque para Azure           | **Sênior**, embora número menor de respondentes |
| **3**   | Analistas técnicos com uso variado de stack                  | Stack mista, destaque para SQL e Azure                        | Equilibrado, mas com leve maioria **Sênior**    |


In [ ]:
# @title Distribuição de Senioridade por Cluster 
import plotly.express as px

fig = px.histogram(
    df,
    x="cluster_tecnico",
    color="2.g_nivel",
    barmode="group",
    title="Distribuição de Senioridade por Cluster Técnico",
    labels={
        "cluster_tecnico": "Cluster Técnico",
        "2.g_nivel": "Senioridade"
    },
    text_auto=True
)

fig.update_layout(xaxis_title="Cluster", yaxis_title="Contagem")
fig.show()


### Heatmap - Relação entre Tempo e Senioridade.
- mostra discrepâncias entre cargo e experiência real.
- Pode ser usado para identificar gaps ou oportunidades de desenvolvimento itnerno, por exemplo plenos com pouca experiência real.

In [ ]:
# @title Heatmap: Senioridade × Tempo de Experiência Médio por Cluster
import plotly.express as px
import pandas as pd

# Pré-processamento 
df["experiencia_dados_anos"] = df["experiencia_dados_anos"].fillna(0)
tabela_heat = df.groupby(["cluster_tecnico", "2.g_nivel"])["experiencia_dados_anos"].mean().unstack()

# Heatmap com Plotly
fig = px.imshow(
    tabela_heat,
    text_auto=".1f",  # Formato com 1 decimal
    color_continuous_scale="YlGnBu",
    aspect="auto",    # Ajuste automático de aspecto
    title="Média de Experiência em Dados por Cluster e Senioridade",
    labels=dict(x="Senioridade", y="Cluster Técnico", color="Anos de Experiência")
)

# Personalizações adicionais
fig.update_layout(
    margin=dict(l=50, r=50, t=80, b=50),
    xaxis_title="Senioridade",
    yaxis_title="Cluster Técnico",
    coloraxis_colorbar=dict(title="Anos")
)

# Melhora a legibilidade dos textos
fig.update_traces(
    textfont_size=12,
    textfont_color="black"
)

fig.show()

### Perfil dominante por Cluster.
| Cluster | Cargo mais comum                  | Cloud Dominante   | Stack Dominante | Insight Principal                                                          |
| ------- | --------------------------------- | ----------------- | --------------- | -------------------------------------------------------------------------- |
| 0       | Analista de Dados/Data Analyst    | AWS               | SQL             | Perfil analista tradicional com foco em SQL e AWS.                         |
| 1       | Outra Opção                       | AWS               | SQL             | Grupo heterogêneo, menos definido — pode incluir transição de carreira.    |
| 2       | Engenheiro de Dados/Data Engineer | AWS               | SQL             | Perfil mais técnico e experiente. Alta senioridade e tempo de experiência. |
| 3       | Analista de Dados/Data Analyst    | Azure (Microsoft) | SQL             | Perfil corporativo com stack mais voltada ao ecossistema Microsoft.        |



In [ ]:
# @title Perfil Dominante por Cluster Técnico
# Funções auxiliares
import pandas as pd
def mais_comum(df_col):
    return df_col.value_counts().idxmax() if not df_col.isnull().all() else "N/A"

# Análise
resumo_clusters = []

for cluster in sorted(df["cluster_tecnico"].unique()):
    df_cluster = df[df["cluster_tecnico"] == cluster]

    cargo_mais_comum = mais_comum(df_cluster["2.f_cargo_atual"])
    cloud_mais_usada = mais_comum(
        df_cluster[[
            "4.h.1_Amazon Web Services (AWS)",
            "4.h.2_Google Cloud (GCP)",
            "4.h.3_Azure (Microsoft)"
        ]].idxmax(axis=1)
    )
    linguagem_mais_usada = mais_comum(
        df_cluster[[
            "4.d.1_SQL",
            "4.d.2_R",
            "4.d.3_Python",
            "4.d.6_Java",
            "4.d.14_JavaScript"
        ]].idxmax(axis=1)
    )

    resumo_clusters.append({
        "Cluster": cluster,
        "Cargo mais comum": cargo_mais_comum,
        "Cloud dominante": cloud_mais_usada.split("_")[-1],
        "Stack dominante": linguagem_mais_usada.split("_")[-1],
        "Quantidade": df_cluster.shape[0]
    })

# Exibir resultado
pd.DataFrame(resumo_clusters)


C:\Users\Tiago\AppData\Local\Temp\ipykernel_31084\1546988363.py:18: FutureWarning:

The behavior of DataFrame.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError

C:\Users\Tiago\AppData\Local\Temp\ipykernel_31084\1546988363.py:27: FutureWarning:

The behavior of DataFrame.idxmax with all-NA values, or any-NA and skipna=False, is deprecated. In a future version this will raise ValueError



Cluster                                  Cargo mais comum  \
0        0                    Analista de Dados/Data Analyst   
1        1                                       Outra Opção   
2        2  Engenheiro de Dados/Data Engineer/Data Architect   
3        3                    Analista de Dados/Data Analyst   

             Cloud dominante Stack dominante  Quantidade  
0  Amazon Web Services (AWS)             SQL        2061  
1  Amazon Web Services (AWS)             SQL        1938  
2  Amazon Web Services (AWS)             SQL         223  
3          Azure (Microsoft)             SQL         995

In [33]:
# Verificar os valores únicos reais da coluna de satisfação
df["2.k_satisfeito_atualmente"].dropna().unique()


array([True, False], dtype=object)

In [38]:
df["2.k_satisfeito_atualmente"].value_counts(dropna=False)


2.k_satisfeito_atualmente
True     3336
False    1527
NaN       354
Name: count, dtype: int64

### Avaliação de Modelo Supervisionado.
- A modelagem usou RandomForestClassifier com 5-fold cross-validation.
- O modelo alcançou F1-socore médio de 0.696 com desvio de apenas de 0.02, o que indica consistência e estabilidade entre os folds o que sugere que há padraão nos dados capaz de prever a variável "satisfação atual".
- A variável alvo (2.k_satisfeito_atualmente) foi convertida para 0 (insatisfeito) e 1 (satisfeito), o que possiblita um problema binário claro.

In [ ]:
#  Importações
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
import numpy as np

#  Prepara os dados
df_ml = df[[
    "2.k_satisfeito_atualmente", 
    "experiencia_dados_anos", 
    "experiencia_ti_anos", 
    "4.d.1_SQL", "4.d.2_R", "4.d.3_Python", "4.d.6_Java", "4.d.14_JavaScript",
    "4.h.1_Amazon Web Services (AWS)", "4.h.2_Google Cloud (GCP)", "4.h.3_Azure (Microsoft)"
]].dropna()

print(f" Total de registros válidos para modelagem: {len(df_ml)}")
if len(df_ml) == 0:
    raise ValueError("❌ Nenhum dado válido para modelar. Verifique os dados!")

#  Variável alvo
y = df_ml["2.k_satisfeito_atualmente"].astype(bool).astype(int)

#  Features
X = df_ml.drop(columns=["2.k_satisfeito_atualmente"])

#  Pipeline com normalização e modelo
pipeline = Pipeline([
    ("scaler", StandardScaler()),
    ("modelo", RandomForestClassifier(n_estimators=200, class_weight="balanced", random_state=42))
])

#  Cross-validation
from sklearn.model_selection import cross_val_score

scores = cross_val_score(
    pipeline,
    X, y,
    cv=5,  # 5 folds
    scoring="f1",  # Pode trocar para "accuracy", "recall", "roc_auc", etc.
    n_jobs=-1
)

#  Resultados
print(" F1-score por fold:", np.round(scores, 3))
print(" F1-score médio:", round(scores.mean(), 3))
print(" Desvio padrão:", round(scores.std(), 3))




🧾 Total de registros válidos para modelagem: 1306
🎯 F1-score por fold: [0.688 0.672 0.687 0.7   0.731]
📊 F1-score médio: 0.696
📉 Desvio padrão: 0.02


### Importância das Variáveis
- O objetivo é visualizar o quanto cada variável técnica e de experiência contribui para a predição da satisfação.
- As variáveis mais importantes para a predião de satisfação foram: experiencia_ti_anos, experiencia_dados_anos, uso de Azure e GCP.
- Isso nos mostra que a satisfação dos profissionais está fortemente ligada ao tempo de atuação na área, e em seguida ao ambiente cloud usado no trabalho.
- Linguagens como Python, R e SQL têm peso menor, o que nos mostra que não são fatores decisivos isoladamente para determinar satisfação atualmente. 

In [59]:
import pandas as pd
import plotly.express as px

# Cria DataFrame com as importâncias
importancias = pd.DataFrame({
    "Variável": X.columns,
    "Importância": modelo.feature_importances_
}).sort_values(by="Importância", ascending=True)  # ordem crescente para barra horizontal

# Gráfico de barras horizontais
fig = px.bar(
    importancias,
    x="Importância",
    y="Variável",
    orientation="h",
    title="📌 Importância das Variáveis para Predição de Satisfação",
    labels={"Importância": "Peso na decisão do modelo", "Variável": "Variável"},
    color="Importância",
    color_continuous_scale="viridis"
)

# Estética adicional
fig.update_layout(
    xaxis_title="Importância (Gini)",
    yaxis_title="",
    coloraxis_showscale=False,
    height=400 + 20 * len(importancias),  # altura dinâmica
    margin=dict(l=100, r=20, t=60, b=20)
)

fig.show()
